temporal-sdk/src/test/java/io/temporal/internal/worker/WorkflowSlotsSmallSizeTests.java

```diff
@@ -0,0 +1,285 @@
+import io.temporal.testUtils.CountingSlotSupplier;
+import io.temporal.testing.internal.SDKTestWorkflowRule;
+import io.temporal.worker.MetricsType;
+import io.temporal.worker.WorkerOptions;
+import io.temporal.worker.tuning.ActivitySlotInfo;
+import io.temporal.worker.tuning.CompositeTuner;
+import io.temporal.worker.tuning.LocalActivitySlotInfo;
+import io.temporal.worker.tuning.WorkflowSlotInfo;
+import io.temporal.workflow.*;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class WorkflowSlotsSmallSizeTests {
+  private final int MAX_CONCURRENT_WORKFLOW_TASK_EXECUTION_SIZE = 2;
+  private final int MAX_CONCURRENT_ACTIVITY_EXECUTION_SIZE = 2;
+  private final int MAX_CONCURRENT_LOCAL_ACTIVITY_EXECUTION_SIZE = 2;
+  private final CountingSlotSupplier<WorkflowSlotInfo> workflowTaskSlotSupplier =
+      new CountingSlotSupplier<>(MAX_CONCURRENT_WORKFLOW_TASK_EXECUTION_SIZE);
+  private final CountingSlotSupplier<ActivitySlotInfo> activityTaskSlotSupplier =
+      new CountingSlotSupplier<>(MAX_CONCURRENT_ACTIVITY_EXECUTION_SIZE);
+  private final CountingSlotSupplier<LocalActivitySlotInfo> localActivitySlotSupplier =
+      new CountingSlotSupplier<>(MAX_CONCURRENT_LOCAL_ACTIVITY_EXECUTION_SIZE);
+  private final TestStatsReporter reporter = new TestStatsReporter();
+  static Semaphore parallelSemRunning = new Semaphore(0);
+  static Semaphore parallelSemBlocked = new Semaphore(0);
+
+  Scope metricsScope =
+      new RootScopeBuilder().reporter(reporter).reportEvery(com.uber.m3.util.Duration.ofMillis(1));
+
+  @Parameterized.Parameter public boolean activitiesAreLocal;
+
+  @Parameterized.Parameters()
+  public static Object[] data() {
+    return new Object[][] {{true}, {false}};
+  }
+
+  @Rule
+  public SDKTestWorkflowRule testWorkflowRule =
+      SDKTestWorkflowRule.newBuilder()
+          .setWorkerOptions(
+              WorkerOptions.newBuilder()
+                  .setWorkerTuner(
+                      new CompositeTuner(
+                          workflowTaskSlotSupplier,
+                          activityTaskSlotSupplier,
+                          localActivitySlotSupplier))
+                  .build())
+          .setMetricsScope(metricsScope)
+          .setActivityImplementations(new TestActivitySemaphoreImpl())
+          .setWorkflowTypes(ParallelActivities.class)
+          .setDoNotStart(true)
+          .build();
+
+  @Before
+  public void setup() {
+    reporter.flush();
+    parallelSemRunning = new Semaphore(0);
+    parallelSemBlocked = new Semaphore(0);
+  }
+
+  @After
+  public void tearDown() {
+    testWorkflowRule.getTestEnvironment().close();
+    assertEquals(
```

Comment:
There is some dupe among these tests with this bit, and asserting on the metrics, etc.

Not sure if it's actually worth it or not to dedupe it, though. Welcome any thoughts on that.